### ISSUE TITLE
Add 15 betting odds features to ML expected points model

### ISSUE THREAD

**Issue by alejio** [alejio/fpl-team-picker#38]
## Overview

Integrate betting odds data from fpl-dataset-builder into the ML expected points model to improve prediction accuracy. Betting odds provide unique market-based signals that complement existing historical performance features.

## Context

- **Data source**: `fpl-dataset-builder` now provides `get_raw_betting_odds()` with ~60 fields per fixture
- **Current ML features**: 84 features (65 base + 15 enhanced + 4 penalty/set-piece)
- **Betting odds coverage**: Match outcome odds, over/under 2.5, Asian Handicap, referee, from multiple bookmakers
- **Documentation**: `../fpl-dataset-builder/docs/BETTING_ODDS_INTEGRATION.md`

### Category 1: Implied Probabilities (6 features)

**1. `team_win_probability`**
- **Calculation**: `1 / AvgCH` (home) or `1 / AvgCA` (away), normalized for overround
- **Formula**: `prob / (prob_home + prob_draw + prob_away)`
- **Why powerful**: Market consensus on match outcome, accounts for injuries/form/motivation
- **Data source**: Closing odds averages (AvgCH, AvgCD, AvgCA)

**2. `opponent_win_probability`**
- **Calculation**: Opponent's win probability (inverse of #1)
- **Why powerful**: Opponent strength from market perspective

**3. `draw_probability`**
- **Calculation**: `1 / AvgCD`, normalized for overround
- **Why powerful**: High draw probability = defensive match, fewer attacking returns

**4. `implied_clean_sheet_probability`**
- **Calculation**: `team_win_prob * 0.65 + draw_prob * 0.45`
- **Why powerful**: Better CS predictor than historical stats (empirical PL data: 65% wins have CS, 45% draws)
- **Position benefit**: GKP/DEF (clean sheet points)

**5. `implied_total_goals`**
- **Calculation**: `2.5 * (Avg_over_2_5 / (Avg_over_2_5 + Avg_under_2_5))`
- **Why powerful**: Market expectation for match tempo/goal volume
- **Data source**: Over/Under 2.5 market aggregates

**6. `team_expected_goals`**
- **Calculation**: `implied_total_goals * (team_win_prob * 0.6 + draw_prob * 0.3 + opponent_win_prob * 0.1)`
- **Why powerful**: Team's expected goal share based on match outcome probabilities
- **Position benefit**: FWD/MID (goals/assists)

### Category 2: Market Confidence & Line Movement (4 features)

**7. `market_consensus_strength`**
- **Calculation**: `1 / (MaxCH - AvgCH + MaxCA - AvgCA)` for favorite
- **Why powerful**: Lower bookmaker spread = higher confidence = more predictable
- **Data source**: Max vs Avg closing odds

**8. `odds_movement_team`**
- **Calculation**: Home: `B365CH - B365H`, Away: `B365CA - B365A`
- **Why powerful**: Line movement shows where "sharp money" went, closing odds more accurate
- **Interpretation**: Positive = odds drifted (less confident), negative = sharpened (more confident)

**9. `odds_movement_magnitude`**
- **Calculation**: `abs(B365CH - B365H) + abs(B365CD - B365D) + abs(B365CA - B365A)`
- **Why powerful**: Large movements indicate new information (injury, suspension, weather)

**10. `favorite_status`**
- **Calculation**: Binary: `1` if `team_win_prob > opponent_win_prob`, else `0`
- **Why powerful**: Favorites have higher floor (defensive points), underdogs higher ceiling (upset wins)

### Category 3: Asian Handicap Insights (3 features)

**11. `asian_handicap_line`**
- **Calculation**: Direct from `AHh` field (e.g., -1.5 = expected to win by 2+ goals)
- **Why powerful**: More granular than win probability, shows expected dominance
- **Data source**: Asian Handicap market

**12. `handicap_team_odds`**
- **Calculation**: Home: `AvgAHH`, Away: `AvgAHA`
- **Why powerful**: Market's confidence in covering the spread

**13. `expected_goal_difference`**
- **Calculation**: `AHh * 0.7` (empirical conversion from handicap to goals)
- **Why powerful**: Attacking/defensive expectation
- **Position benefit**: All positions (attacking returns for positive, defensive for negative)

### Category 4: Match Context (2 features)

**14. `over_under_signal`**
- **Calculation**: `(Avg_over_2_5 - Avg_under_2_5) / (Avg_over_2_5 + Avg_under_2_5)`
- **Range**: -1 (defensive match) to +1 (high-scoring match)
- **Why powerful**: Match tempo expectation affects all positions

**15. `referee_encoded`**
- **Calculation**: LabelEncoder on referee names
- **Why powerful**: Referee tendencies (card frequency, penalty leniency) affect points
- **Note**: Full value requires historical referee stats database (future enhancement)

## Temporal Validation (Leak-Free Guarantee)

✅ **All features are leak-free**:
- Closing odds available at kickoff (BEFORE points determined)
- Pre-match odds available 1-2 weeks before match
- NO match statistics used (those are post-match outcomes)
- Historical player stats still use `.shift(1)` as usual

**Key difference from performance features**:
- Betting odds are FORWARD-LOOKING predictions
- Can be used for GW N to predict GW N (no shift needed)
- Unlike player stats which require shift(1) to avoid leakage

### Position-Specific Benefits
- **GKP/DEF**: `implied_clean_sheet_probability`, `implied_total_goals` (inverse), `expected_goal_difference`
- **MID**: All features (balanced scoring profile)
- **FWD**: `team_expected_goals`, `implied_total_goals`, `favorite_status`

### Estimated Performance Improvement
- **MAE reduction**: 5-8% (based on sports analytics research)
- **Biggest gains**: GW1-5 (sparse historical data, odds provide unique signal)
- **Edge cases**: New signings, manager changes, fixture congestion (market reacts faster than historical stats)

### Academic Validation
- 20+ studies show closing odds are most accurate pre-match outcome predictor
- Odds movement captures information asymmetry
- Asian Handicap provides better signal than 1X2 for scoring margins

### Phase 1: fpl-dataset-builder (Derived Data Layer)

**File**: `client/derived.py`

Create `get_derived_betting_features(gameweek: int = None) -> pd.DataFrame`:
- Calculate all 15 features at fixture level
- Expand to player level (separate rows for home/away team players)
- Handle missing odds gracefully (neutral defaults)
- Return columns:
  - `gameweek`, `fixture_id`, `player_id`, `is_home`
  - All 15 feature columns
  - `as_of_utc` (metadata timestamp)

**Neutral defaults for missing odds** (GW1 or incomplete bookmaker data):
```python
{
    'team_win_probability': 0.33,  # neutral 3-way split
    'opponent_win_probability': 0.33,
    'draw_probability': 0.33,
    'implied_clean_sheet_probability': 0.35,  # PL average
    'implied_total_goals': 2.5,  # PL average
    'team_expected_goals': 1.25,  # neutral split
    'market_consensus_strength': 0.5,  # unknown
    'odds_movement_team': 0.0,  # no movement
    'odds_movement_magnitude': 0.0,
    'favorite_status': 0.5,  # unknown
    'asian_handicap_line': 0.0,  # neutral handicap
    'handicap_team_odds': 2.0,  # neutral
    'expected_goal_difference': 0.0,
    'over_under_signal': 0.0,  # neutral
    'referee_encoded': -1,  # unknown/new referee
}
```

**Testing**:
- Unit tests: Probability normalization, handicap conversion, missing data handling
- Integration test: Load GW1-10 data, verify all features non-null

### Phase 2: fpl-team-picker (ML Feature Engineering)

**File**: `fpl_team_picker/domain/services/ml_feature_engineering.py`

**Changes to `FPLFeatureEngineer`**:

1. **Constructor update**:
```python
def __init__(
    self,
    fixtures_df: pd.DataFrame,
    teams_df: pd.DataFrame,
    team_strengths: Union[Dict[str, float], Dict[int, Dict[str, float]]],
    betting_features_df: Optional[pd.DataFrame] = None,  # NEW
    ownership_trends_df: Optional[pd.DataFrame] = None,
    value_analysis_df: Optional[pd.DataFrame] = None,
    fixture_difficulty_df: Optional[pd.DataFrame] = None,
    raw_players_df: Optional[pd.DataFrame] = None,
):
    # ...
    self.betting_features_df = betting_features_df
```

2. **New method**:
```python
def _add_betting_odds_features(self, df: pd.DataFrame) -> pd.DataFrame:
    """Add 15 betting odds features with temporal validation."""
    if self.betting_features_df is None:
        # Graceful degradation: fill with neutral defaults
        for feature, default in BETTING_ODDS_DEFAULTS.items():
            df[feature] = default
        return df
    
    # NO SHIFT NEEDED - odds are forward-looking (available pre-match)
    # Unlike player stats, odds for GW N predict GW N (no data leakage)
    
    # Merge betting features into player-level data
    df = df.merge(
        self.betting_features_df,
        on=['gameweek', 'player_id'],  # or ['gameweek', 'fixture_id', 'is_home']
        how='left'
    )
    
    # Fill missing (GW1 or missing odds) with neutral values
    df.fillna(BETTING_ODDS_DEFAULTS, inplace=True)
    
    return df
```

3. **Update `fit_transform()` method**:
- Call `_add_betting_odds_features()` after `_add_fixture_features()`

4. **Update `_get_feature_columns()`**:
- Add 15 betting odds feature names to list

**Testing**:
- Integration test: Load GW1-10 data with betting features, verify 99 total features (84 + 15)
- Temporal validation test: Verify no future data leakage
- Missing data test: Verify graceful handling when betting_features_df is None

### Phase 3: ml_xp_notebook.py Integration

**File**: `fpl_team_picker/interfaces/ml_xp_notebook.py`

1. **Load betting features**:
```python
# Load betting features from fpl-dataset-builder
betting_features_df = client.get_derived_betting_features()  # All gameweeks
```

2. **Pass to FPLFeatureEngineer**:
```python
feature_engineer = FPLFeatureEngineer(
    fixtures_df=fixtures_df,
    teams_df=teams_df,
    team_strengths=team_strengths,
    betting_features_df=betting_features_df,  # NEW
    ownership_trends_df=ownership_trends_df,
    # ...
)
```

3. **Update feature count display**:
- Change "84 features" → "99 features (84 base + 15 betting odds)"

4. **Add betting odds analysis section** (optional):
- Feature importance for betting odds features
- Correlation analysis with existing features
- Position-specific feature importance

### Phase 4: Documentation

**Files to update**:

1. **`fpl-team-picker/CLAUDE.md`**:
   - Update "ML Development" section
   - Add "Betting Odds Features (15)" to feature list
   - Update feature count: 84 → 99
   - Add example: `client.get_derived_betting_features()`

2. **`fpl-dataset-builder/CLAUDE.md`**:
   - Add `get_derived_betting_features()` to "Client Methods" section
   - Add usage example

3. **`scripts/README.md`**:
   - Update TPOT feature count: 84 → 99

## Data Availability

**Source**: `fpl-dataset-builder` 
- `get_raw_betting_odds(gameweek)` - ~60 raw fields per fixture
- NEW: `get_derived_betting_features(gameweek)` - 15 calculated features per player-fixture

**Update frequency**: Weekly (Mondays) from football-data.co.uk

**Historical coverage**: 2024-25 season onwards (future: extend to 2023-24, 2022-23)

## Validation Criteria

✅ All 15 features use only pre-match/closing odds (no post-match statistics)
✅ Temporal validation: No data leakage (odds available before points)
✅ Graceful degradation: Missing odds → neutral defaults
✅ Integration tests pass (GW1-10 data)
✅ Feature count: 84 → 99
✅ Documentation updated
✅ No regression in model training time (<10% increase acceptable)

## Success Metrics

- [ ] MAE reduction: Target 5-8% on test set (GW8-10)
- [ ] GW1-5 improvement: Target 10%+ (sparse historical data period)
- [ ] Position-specific gains: GKP/DEF (CS), FWD (goals)
- [ ] Feature importance: Betting odds in top 20 features
- [ ] No temporal leakage: Validation set performance consistent with test set

## Questions / Decisions Needed

1. **Bookmaker preference**: Use closing averages (AvgCH/CD/CA) as primary, or Pinnacle (sharpest)?
   - **Recommendation**: Closing averages (most stable, handles missing bookmakers)

2. **Referee features**: Build historical referee database now or defer?
   - **Recommendation**: Defer to future enhancement (feature #15 uses simple encoding for now)

3. **Missing data threshold**: Exclude fixtures with missing odds from training, or use neutral defaults?
   - **Recommendation**: Use neutral defaults (more training data, graceful degradation)

4. **Feature selection**: All 15 features or prioritize top 10 based on feature importance?
   - **Recommendation**: Start with all 15, prune low-importance features after initial evaluation

## Related Issues

- Depends on: fpl-dataset-builder betting odds integration (completed)
- Relates to: Enhanced feature engineering initiative

---

**Feature Count Summary**: 84 → 99 features (15 new betting odds features)
**Estimated Development Time**: 4-6 hours (2h derived layer + 2h ML integration + 1h testing + 1h docs)
**Risk**: Low (graceful degradation if odds unavailable, no breaking changes)
